### PR TITLE
feat: add `CommutativeSemiGroup`

### LINKED ISSUE #1922: Type Classes: Add CommutativeSemiGroup
Propose a design

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes #1922

I'm not sure how to enforce the commutative law when extending `SemiGroup` which has instances of combine defined for `String` as concatenation which is clearly not commutative. Does the compiler enforce the laws yet? Any ideas? I'll take a look at typelevel / cats meanwhile.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
I think CommutativeSemiGroup should re-export combine.

See https://github.com/flix/flix/issues/2649

**Comment by magnus-madsen** [comment]
I think the TC can be lawful (remove the lawless modifier).

**Comment by jaschdoc** [comment]
What do you think of this? I'm not sure `Validation[t, e]` is commutative since it has this arm
```flix
case (Failure(es1), Failure(es2))   => Failure(Nel.append(es1, es2))
```

**Comment by jaschdoc** [comment]
I still need to add instances to `Int32` (and friends). What other data structures are commutative and associatve?

**Comment by magnus-madsen** [comment]
Strictly fewer than are semi groups, right
It's always a subset, so that's the starting point.

**Comment by magnus-madsen** [comment]
Validation is not commutative because it contains a list and lists have an order.

File: main/src/library/CommutativeSemiGroup.flix [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    pub def combineN(x: a, n: Int32): a = SemiGroup.combineN(x, n)
+
+    law commutative: forall(x: a, y: a) with Eq[a] . CommutativeSemiGroup.combine(x, y) == CommutativeCommutativeSemiGroup.combine(y, x)
+
+}
+
+instance CommutativeSemiGroup[Unit] {
+}
+
+instance CommutativeSemiGroup[Option[a]] with CommutativeSemiGroup[a] {
```
Comment (magnus-madsen) [inline main/src/library/CommutativeSemiGroup.flix:40]: - Instances for enums (e.g. Option) belong in their namespace files. (This should also be changed for plain SemiGroup).
- The {} braces can be dropped.

File: main/src/library/CommutativeSemiGroup.flix [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ */
+
+
+///
+/// A type class for types that form a commutative semigroup.
+///
+pub class CommutativeSemiGroup[a] with SemiGroup[a] {
+
+    ///
+    /// An associative binary operation on `a`.
```
Comment (magnus-madsen) [inline main/src/library/CommutativeSemiGroup.flix:24]: and commutative :)

Comment (jaschdoc) [reply main/src/library/CommutativeSemiGroup.flix:24]: Yup coming right up :)

**Comment by jaschdoc** [comment]
Should I move the `SemiGroup` instance for `String` to its respective namespace or keep it in `SemiGroup`?

**Comment by magnus-madsen** [comment]
> 

For the base types, which String is, we keep in the TC namespace.

**Comment by jaschdoc** [comment]
How come we don't have any TC instances for integers? Are these assumed to be commutative?

**Comment by magnus-madsen** [comment]
> How come we don't have any TC instances for integers? Are these assumed to be commutative?

Ignoring overflow, I think integers under addition form a semi group and commutative semi group.

**Comment by jaschdoc** [comment]
> I think integers under addition form a semi group and commutative semi group.
Hmm yeah but I think ℤ under multiplication also forms both structures. Do we just default to addition?

**Comment by magnus-madsen** [comment]
> > I think integers under addition form a semi group and commutative semi group.
> 
> Hmm yeah but I think ℤ under multiplication also forms both structures. Do we just default to addition?

Lets default to addition, but we can add them in a separate PR?

**Comment by jaschdoc** [comment]
Sounds good

**Comment by jaschdoc** [comment]
Ready for review

**Pull request merged**